File: js/packages/app/src/__tests__/features/object-references/external/wrappers/object-references-example/src/polywrap/external/module/InternalWrapInstance.ts
import { IExternalWrapInstance, IInternalWrapInstance } from "@nerfzael/reim-wrap-as";
import { invoke as invokeGlobalFunction } from "../../internal/global-functions/invokeGlobalFunction";
import { invokeClassMethod } from "../../internal/classes/invokeClassMethod";
import { InternalResource } from "../../dt/InternalResource";
import { ExternalResource } from "../../dt/ExternalResource";

export class InternalWrapInstance extends IInternalWrapInstance {
  constructor() {
    super();
  }

  invokeResource(resource: u32, buffer: ArrayBuffer, externalWrapInstance: IExternalWrapInstance): ArrayBuffer {
    switch (resource) {
      case ExternalResource.InvokeGlobalFunction:
        return invokeGlobalFunction(buffer, externalWrapInstance);
      case InternalResource.InvokeClassMethod: 
        return invokeClassMethod(buffer, externalWrapInstance);
      default:
        throw new Error("Unknown resource");
    }
  }
}

File: rust/app/src/wrappers/external_references/src/polywrap/external/module/internal_wrap_instance.rs
use std::sync::Arc;

use async_trait::async_trait;
use reim_wrap::{ ExternalModule, InternalModule };
use crate::polywrap::internal::global_functions::{ invoke as invoke_global_function };
use crate::polywrap::internal::classes::{ invoke_class_method };
use crate::polywrap::dt::{ InternalResource };

struct InternalWrapModule {}

impl InternalWrapModule {
    pub fn new() -> Self {
        InternalWrapModule {}
    }
}

#[async_trait]
impl InternalModule for InternalWrapModule {
  async fn invoke_resource(&self, resource: u32, buffer: &[u8], external_module: Arc<dyn ExternalModule>) -> Vec<u8> {
    match resource {
      x if x == InternalResource::InvokeGlobalFunction as u32 => invoke_global_function(buffer, external_module),
      x if x == InternalResource::InvokeClassMethod as u32 => invoke_class_method(buffer, external_module),
      _ => panic!("Unknown resource: {}", resource.to_string()),
    }
  }
}
